src/main/java/com/metamx/common/parsers/JavascriptParser.java

```diff
@@ -0,0 +1,154 @@
+
+/**
+ */
+public class JavascriptParser implements Parser<String, Object>
+{
+  private static final Logger log = new Logger(JavascriptParser.class);
+
+  private static Function<Object, Object> compile(String function)
+  {
+    final ContextFactory contextFactory = ContextFactory.getGlobal();
+    final Context context = contextFactory.enterContext();
+    context.setOptimizationLevel(9);
+
+    final ScriptableObject scope = context.initStandardObjects();
+
+    final org.mozilla.javascript.Function fn = context.compileFunction(scope, function, "fn", 1, null);
+    Context.exit();
+
+    return new Function<Object, Object>()
+    {
+      public Object apply(Object input)
+      {
+        // ideally we need a close() function to discard the context once it is not used anymore
+        Context cx = Context.getCurrentContext();
+        if (cx == null) {
+          cx = contextFactory.enterContext();
+        }
+
+        final Object res = fn.call(cx, scope, scope, new Object[]{input});
+        return res != null ? Context.toObject(res, scope) : null;
+      }
+    };
+  }
+
+  private final Function<Object, Object> fn;
+  private final String listDelimiter;
+  private final Splitter listSplitter;
+  private final Function<String, Object> valueFunction;
+
+  private ArrayList<String> fieldNames = null;
+
+  public JavascriptParser(
+      final String function,
+      final Optional<String> listDelimiter
+  )
+  {
+    this.fn = compile(function);
+    this.listDelimiter = listDelimiter.isPresent() ? listDelimiter.get() : Parsers.DEFAULT_LIST_DELIMITER;
+    this.listSplitter = Splitter.on(this.listDelimiter);
+    this.valueFunction = new Function<String, Object>()
+    {
+      @Override
+      public Object apply(String input)
+      {
+        if (input.contains(JavascriptParser.this.listDelimiter)) {
+          return Lists.newArrayList(
+              Iterables.transform(
+                  listSplitter.split(input),
+                  ParserUtils.nullEmptyStringFunction
+              )
+          );
+        } else {
+          return ParserUtils.nullEmptyStringFunction.apply(input);
+        }
+      }
+    };
+  }
+
+  public JavascriptParser(
+      final String function,
+      final Optional<String> listDelimiter,
+      final Iterable<String> fieldNames
+  )
+  {
+    this(function, listDelimiter);
+
+    setFieldNames(fieldNames);
+  }
+
+  public Function<Object, Object> getFn()
+  {
+    return fn;
+  }
+
+  @Override
+  public Map<String, Object> parse(String input)
+  {
+    try {
+      Iterable values;
+      Object compiled = fn.apply(input);
+      if (compiled instanceof Iterable) {
+        values = (Iterable) compiled;
+      } else {
+        values = Arrays.asList(compiled.toString());
+      }
+
+      if (fieldNames == null) {
+        setFieldNames(ParserUtils.generateFieldNames(Iterables.size(values)));
+      }
+
+      return Utils.zipMapPartial(fieldNames, Iterables.transform(values, valueFunction));
```

Comment:
I think for a JavaScript parser it'd be more natural to return an object, rather than an array of things that gets zipped with fieldNames. So IMO fieldNames should be ignored and the JavaScript function should be expected to return an object.